src/PostgREST/App.hs

```diff
@@ -205,8 +204,36 @@ app dbStructure proc conf apiRequest =
                     then responseLBS s [toHeader contentType, r] (toS body)
                     else responseLBS s [r] ""
 
-        (ActionDelete, TargetIdent _, Nothing) ->
-          case mutateSqlParts of
+        (ActionSingleUpsert, TargetIdent (QualifiedIdentifier tSchema tName), Just (PayloadJSON payload pType pKeys)) ->
+          case mutateSqlParts tSchema tName of
+            Left errorResponse -> return errorResponse
+            Right (sq, mq) -> do
+              let isSingle = case pType of
+                               PJArray len -> len == 1
+                               PJObject -> True
+                  colNames = colName <$> tableCols dbStructure tSchema tName
+              if topLevelRange /= allRange
+                then return $ simpleError status400 [] "Range header and limit/offset querystring parameters are not allowed for PUT"
+              else if not isSingle
+                then return $ simpleError status405 [] "PUT payload must contain a single row"
+              else if S.fromList colNames /= pKeys
+                then return $ simpleError status405 [] "You must specify all columns in the payload when using PUT"
+              else do
+                row <- H.query (toS payload) $
+                       createWriteStatement sq mq (contentType == CTSingularJSON) False
+                                            (contentType == CTTextCSV) (iPreferRepresentation apiRequest) []
+                let (_, queryTotal, _, body) = extractQueryResult row
+                if queryTotal /= 1
```

Comment:
Nice, `queryTotal /= 1` is a clever way to detect that the PUT is being applied incorrectly, rather than comparing the filters in the request to primary keys etc etc.

Reply:
Looking further into the code, I see that `mutateRequest` checks for `ActionSingleUpsert` that the request contains equality filters on all primary keys. Does that make the `queryTotal /= 1` here unnecessary, or what is the interaction between the logic?

Reply:
Is not clear from the code but I'm enforcing a different condition here, so first of all my interpretation of [RFC 7231](https://tools.ietf.org/html/rfc7231#section-4.3.4) is that the PUT uri must identify the payload being inserted so that the next GET can return the same exact payload right? So what I'm doing here is this:

```haskell
-- table numbers is empty
it "fails if the uri doesn't identify the payload" $
  put "/numbers?id=eq.7" [str| [ { "id": 11, "name": "eleven"} ]|] `shouldRespondWith` 405
```

This is actually done by the [sql query](https://github.com/steve-chavez/postgrest/blob/upsert/src/PostgREST/QueryBuilder.hs#L290) that WHEREs the payload according to the querystring parameters, if that WHERE gives false then nothing gets upserted and the `queryTotal` will be 0.

Reply:
I tried to not have the restriction of just allowing primary key cols `eq` filters first, but without that there's no reliable way of identifying the payload with the same sql query,  consider:
```http
-- id is persons pk, there's already many records with the name John
PUT /persons?name=eq.John
{ "id": 12, "first_name": "John", "last_name": "Doe" }
```
The WHERE would give true and the PUT would succeed with a `queryTotal == 1` but since name is not the pk subsequent `GET /persons` would have many records.

Reply:
@begriffs Let me know if I'm interpreting the RFC right and if so how can this logic be made more simple/clear.